galaxyproject/galaxy
# Add some Zarr-based datatypes

Requires #17614 and #19077

Includes the following datatypes:

### General Zarr datatypes
- `CompressedZarrZipArchive` (`zarr.zip`): represents a Zarr ZipStore. It seems to have some limitations (i.e. it doesn't work with zips containing the zarr store in a subfolder).
- `ZarrDirectory` (`zarr`): represents a Zarr DirectoryStore. Contains the zarr structure in the `extra_files_path` of the dataset. I wonder if there is a way to make the "main" dataset to really point to the actual root folder in extra_files like a symlink or something like this, but I don't know if this really makes sense. Right now, to access your zarr store root you will need to use the input like this `'$zarrinput.extra_files_path/$zarrinput.metadata.store_root'` (see `Handling different input types` section below)

### OME-Zarr datatypes
- `CompressedOMEZarrZipArchive` (`ome_zarr.zip`): Similar to CompressedZarrZipArchive but expects to find an `OME/METADATA.ome.xml` file in the store root so it can be easily converted/extracted to an `OMEZarr` directory.
- `OMEZarr` (`ome_zarr`): Similar to ZarrDirectory but identify this datatype as an OME Zarr image.

### Handling different input types
If your zarr input is defined like this:
```xml
<inputs>
    <param name="zarrinput" type="data" format="zarr,zarr.zip" allow_uri_if_protocol="https" label="Zarr input"/>
</inputs>
```
You need to access the zarr store in a "slightly" different way depending on how the store is provided as input.

#### Input as `zarr.zip`


https://github.com/user-attachments/assets/b37ca798-d895-405a-8644-5d127a5a50fb



You can upload a zip file containing the zarr store. In order to open it in your tool you need to do the following:

```python
if '$zarrinput.extension' == 'zarr.zip':
    compression = int('$zarrinput.metadata.compression' or 0)
    zip_store = zarr.ZipStore('$zarrinput', mode='r', compression=compression)
    input_zarr = zarr.open(zip_store, mode='r')`
```
Please note that the root store must be in the root of the zip (i.e. no subdirectory containing the store)

#### Input as `zarr` (Directory)
You cannot directly upload a directory, but you can upload a zarr.zip and then extract it to a "zarr directory dataset".


https://github.com/user-attachments/assets/3d4f3975-5b42-4b12-99d4-9108bf0105ee



To open this kind of zarr in your tool you must do the following:

```python
if '$zarrinput.extension' == 'zarr' and '$zarrinput.state' != 'deferred':
        input_store = '$zarrinput.extra_files_path/$zarrinput.metadata.store_root'
        input_zarr = zarr.open(input_store, mode='r')`
```
Note that you need to reference the extra_files_path and append the `$zarrinput.metadata.store_root` to it. This will point to the right internal directory where the zarr store has been extracted and will work even if the uploaded zip had a parent subfolder containing the zarr store.


#### Input as a `deferred zarr` (URI to a remote zarr)

https://github.com/user-attachments/assets/8dca0155-6948-4c32-a090-07a4962fff59



In your tool, you can just directly use '$zarrinput', it will point to the remote URI. But you need to check this is actually a deferred dataset like this:

```python
if '$zarrinput.extension' == 'zarr' and '$zarrinput.state' == 'deferred':
        input_zarr = zarr.open('$zarrinput', mode='r')`
```

## How to test the changes?
  - For testing the "directory" variant datatype:
    - Upload any zarr directory in a zip file. You can download them from any public repository or generate one using https://zarr.readthedocs.io/en/stable/tutorial.html
    - Go to `Edit Attributes` then select the `Datatypes` section and convert it to `zarr`.
  - For testing the "deferred" variant (remote URI):
    - Paste the following URI as content: `https://uk1s3.embassy.ebi.ac.uk/idr/zarr/v0.4/idr0062A/6001240.zarr`. You can find more examples [here](https://www.openmicroscopy.org/2020/11/04/zarr-data.html).
    - Make sure you set the dataset options to `Defer dataset resolution`
  - You can use this "simple" tool wrapper to test or make your own:
 
```xml
<tool id="zarr_test" name="Zarr Test" version="0.1.0+galaxy0" profile="23.0">
    <description>test wrapper for zarr format</description>
    <requirements>
        <requirement type="package" version="2.18.3">zarr</requirement>
        <requirement type="package" version="2024.9.0">s3fs</requirement>
    </requirements>
    <command detect_errors="exit_code"><![CDATA[
        python $script
    ]]></command>
    <configfiles>
        <configfile name="script"><![CDATA[
import zarr

input_store = None
if '$zarrinput.extension' == 'zarr':
    if '$zarrinput.is_deferred':
        print('Using remote S3 store as input')
        # Deferred datasets will return directly the URI in combination with allow_uri_if_protocol
        input_store = '$zarrinput'
    else:
        print('Using local directory store as input')
        # TODO: Is there a way for the dataset file to directly reference the extra_files_path?
        # We need to append the store_root to the extra_files_path to get the full path to the zarr store
        # in case it was extracted to a subdirectory
        input_store = '$zarrinput.extra_files_path/$zarrinput.metadata.store_root'
elif '$zarrinput.extension' == 'zarr.zip':
    compression = int('$zarrinput.metadata.compression' or 0)
    print(f'Using zipped store as input with compression {compression}')
    # Currently only zip files containing the zarr store at the root (i.e. no subdirectories) are supported
    input_store = zarr.ZipStore('$zarrinput', mode='r', compression=compression)
else:
    raise ValueError('Unsupported input format')

if input_store is None:
    raise ValueError('Unable to determine input store. Your input dataset may be an unsupported format')

# Inspect the input zarr
input_zarr = zarr.open(input_store, mode='r')
print("Input zarr info:")
print(f"  {input_zarr.info_items()}")

# Create the output store where the new zarr will be written
output_store = '$zarroutput.extra_files_path'
output_zarr = zarr.open(output_store, mode='w')
foo = output_zarr.create_group('foo')
bar = foo.create_group('bar')
baz = bar.zeros('baz', shape=(10000, 10000), chunks=(1000, 1000), dtype='i4')
        ]]></configfile>
    </configfiles>
    <inputs>
        <param name="zarrinput" type="data" format="zarr,zarr.zip" allow_uri_if_protocol="https" label="Zarr input"/>
    </inputs>
    <outputs>
        <data name="zarroutput" format="zarr" label="${tool.name} on ${on_string}: New ZARR"/>
    </outputs>
    <help><![CDATA[

simple wrapper to test zarr datatype

    ]]></help>
</tool>
```


## License
